File: src/main/java/me/zeroeightsix/kami/gui/mc/KamiGuiUpdateNotification.java
package me.zeroeightsix.kami.gui.mc;

import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.util.WebUtils;
import net.minecraft.client.gui.*;

import java.net.URI;
import java.net.URISyntaxException;

import static me.zeroeightsix.kami.KamiMod.WEBSITE_LINK;

/**
 * Created by Dewy on 09/04/2020
 */
public class KamiGuiUpdateNotification extends GuiScreen {

    private final String title;
    private final String message;
    private final int singleOrMulti;

    public KamiGuiUpdateNotification(String title, String message, int singleOrMulti) {
        super();

        this.title = title;
        this.message = message;
        this.singleOrMulti = singleOrMulti;
    }

    @Override
    public void initGui() {
        super.initGui();

        this.buttonList.add(new GuiButton(0, this.width / 2 - 100, 200, "Download Latest (Recommended)"));
        this.buttonList.add(new GuiButton(1, this.width / 2 - 100, 230, KamiMod.colour + "cUse Current Version"));
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        drawGradientRect(0, 0, this.width, this.height, -12574688, -11530224);

        drawCenteredString(this.fontRenderer, this.title, this.width / 2, 80, 10260478);
        drawCenteredString(this.fontRenderer, this.message, this.width / 2, 110, 16777215);

        super.drawScreen(mouseX, mouseY, partialTicks);
    }

    @Override
    protected void keyTyped(char typedChar, int keyCode) {}

    @Override
    protected void actionPerformed(GuiButton button) {

        if (button.id == 0) {
            try {
                WebUtils.openWebLink(new URI(WEBSITE_LINK + "/download"));

                if (singleOrMulti == 1) {
                    mc.displayGuiScreen(new GuiWorldSelection(new GuiMainMenu()));

                    return;
                }

                mc.displayGuiScreen(new GuiMultiplayer(new GuiMainMenu())); // Multi
            } catch (URISyntaxException e) {
                KamiMod.log.error("Contact the KAMI Blue developers. Download link could not be parsed into URI reference form.");
            }

            return;
        }

        if (singleOrMulti == 1) { // Single
            mc.displayGuiScreen(new GuiWorldSelection(new GuiMainMenu()));

            return;
        }

        mc.displayGuiScreen(new GuiMultiplayer(new GuiMainMenu())); // Multi
    }
}

File: src/main/java/me/zeroeightsix/kami/gui/mc/KamiGuiUpdateNotification.kt
package me.zeroeightsix.kami.gui.mc

import com.google.gson.JsonParser
import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.util.WebUtils
import me.zeroeightsix.kami.util.color.ColorConverter
import net.minecraft.client.gui.*
import net.minecraft.util.text.TextFormatting
import org.kamiblue.commons.utils.ConnectionUtils
import java.io.IOException

class KamiGuiUpdateNotification(private val buttonId: Int) : GuiScreen() {

    private val message = "A newer release of KAMI Blue is available ($latest)."

    override fun initGui() {
        super.initGui()
        buttonList.add(GuiButton(0, width / 2 - 100, 200, "Download Latest (Recommended)"))
        buttonList.add(GuiButton(1, width / 2 - 100, 230, "${TextFormatting.RED}Use Current Version"))
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        drawDefaultBackground()
        drawCenteredString(fontRenderer, title, width / 2, 80, ColorConverter.rgbToHex(155, 144, 255))
        drawCenteredString(fontRenderer, message, width / 2, 110, ColorConverter.rgbToHex(255, 255, 255))

        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    override fun actionPerformed(button: GuiButton) {
        if (button.id == 0) WebUtils.openWebLink(KamiMod.WEBSITE_LINK + "/download")

        val screen = if (buttonId == 1) GuiWorldSelection(GuiMainMenu()) // Single
        else GuiMultiplayer(GuiMainMenu()) // Multi

        mc.displayGuiScreen(screen)
    }

    companion object {
        private const val title = "KAMI Blue Update"

        var latest: String? = null // latest version (null if no internet or exception occurred)
        var isLatest = false

        @JvmStatic
        fun updateCheck() {
            try {
                KamiMod.LOG.info("Attempting KAMI Blue update check...")

                val parser = JsonParser()
                val rawJson = ConnectionUtils.requestRawJsonFrom(KamiMod.DOWNLOADS_API) {
                    throw it
                }

                latest = parser.parse(rawJson).asJsonObject.getAsJsonObject("stable")["name"].asString
                isLatest = latest.equals(KamiMod.VER_STABLE)

                if (!isLatest) {
                    KamiMod.LOG.warn("You are running an outdated version of KAMI Blue.\nCurrent: ${KamiMod.VER_STABLE}\nLatest: $latest")
                } else {
                    KamiMod.LOG.info("Your KAMI Blue (" + KamiMod.VER_STABLE + ") is up-to-date with the latest stable release.")
                }
            } catch (e: IOException) {
                KamiMod.LOG.error("Oes noes! An exception was thrown during the update check.", e)
            }
        }
    }
}